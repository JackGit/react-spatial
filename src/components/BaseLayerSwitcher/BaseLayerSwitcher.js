/* eslint-disable jsx-a11y/interactive-supports-focus */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FaChevronCircleLeft } from 'react-icons/fa';
import Layer from '../../layers/Layer';
import img from '../../images/baselayer/osm.baselayer.png';

import './BaseLayerSwitcher.scss';

const propTypes = {
  /**
   * An array of react-spatial layers.
   */
  layers: PropTypes.arrayOf(PropTypes.instanceOf(Layer)).isRequired,

  /**
   * Object containing relative paths to the base layer images.
   */
  layerImages: PropTypes.objectOf(PropTypes.string),

  /**
   * CSS class to apply on the container.
   */
  className: PropTypes.string,

  /**
   * Alternative text rendered if layer images can't be loaded
   */
  altText: PropTypes.string,

  /**
   * Button titles.
   */
  titles: PropTypes.shape({
    button: PropTypes.string,
    openSwitcher: PropTypes.string,
    closeSwitcher: PropTypes.string,
  }),
};

const defaultProps = {
  className: 'rs-base-layer-switcher',
  altText: 'Source not found',
  titles: {
    button: 'Base layers',
    openSwitcher: 'Open Baselayer-Switcher',
    closeSwitcher: 'Close Baselayer-Switcher',
  },
  layerImages: undefined,
};

const getVisibleLayer = layers => {
  return layers.find(layer => layer.getVisible());
};

const getNextImage = (currentLayer, layers, layerImages) => {
  const currentIndex = layers.indexOf(
    layers.find(layer => layer === currentLayer),
  );
  const nextIndex = currentIndex + 1 === layers.length ? 0 : currentIndex + 1;
  return layerImages[nextIndex];
};

function BaseLayerSwitcher({
  layers,
  layerImages,
  altText,
  className,
  titles,
}) {
  const baseLayers = layers.filter(layer => layer.getIsBaseLayer());
  const [switcherOpen, setSwitcherOpen] = useState(false);
  const [isClosed, setIsClosed] = useState(true);
  const [currentLayer, setCurrentLayer] = useState(getVisibleLayer(baseLayers));

  /* Images are loaded from props if provided, fallback from layer */
  const images = layerImages
    ? Object.keys(layerImages).map(layerImage => layerImages[layerImage])
    : baseLayers.map(layer => layer.previewImage);

  const openClass = switcherOpen ? ' rs-open' : '';
  const closedClass = isClosed ? ' rs-closed' : '';

  const onButtonSelect = (open, layer) => {
    if (!open) {
      setSwitcherOpen(true);
      return;
    }
    setCurrentLayer(layer);
    layer.setVisible(true);
    setSwitcherOpen(false);
  };

  const setTabIndex = (closed, buttonIndex) => {
    if (closed) {
      return buttonIndex === 0 ? '0' : '-1';
    }
    return '0';
  };

  if (!baseLayers || baseLayers.length < 2) {
    return null;
  }

  useEffect(() => {
    /* Ensure correct layer is active on app load */
    if (currentLayer !== getVisibleLayer(baseLayers)) {
      setCurrentLayer(getVisibleLayer(baseLayers));
    }
  }, [currentLayer]);

  useEffect(() => {
    /* Used for correct layer image render with animation */
    let timeout;
    if (!switcherOpen) {
      timeout = setTimeout(() => {
        setIsClosed(true);
      }, 200);
    } else {
      setIsClosed(false);
    }
    return () => clearTimeout(timeout);
  }, [switcherOpen]);

  const toggleBtn =
    baseLayers.length > 0 ? (
      <div
        className="rs-base-layer-switcher-button rs-close"
        role="button"
        onClick={() => setSwitcherOpen(false)}
        onKeyPress={e => e.which === 13 && setSwitcherOpen(false)}
        tabIndex="0"
        aria-label={titles.closeSwitcher}
        title={titles.closeSwitcher}
      >
        <FaChevronCircleLeft size={15} focusable={false} />
      </div>
    ) : null;

  return (
    <div className={`${className}${openClass}`}>
      <div className="rs-base-layer-switcher-layers">
        {!isClosed && toggleBtn}
        {baseLayers.map((layer, index) => {
          const layerName = layer.getName();
          const buttonTitle = isClosed ? titles.openSwitcher : layerName;
          const activeClass =
            layerName === currentLayer.getName() ? ' rs-active' : '';
          return (
            <div
              key={layer.key}
              className={`rs-base-layer-switcher-button rs-layer${openClass}${activeClass}`}
              role="button"
              title={buttonTitle}
              aria-label={buttonTitle}
              onClick={() => onButtonSelect(switcherOpen, layer)}
              onKeyPress={e => {
                if (e.which === 13) {
                  onButtonSelect(switcherOpen, layer);
                }
              }}
              tabIndex={setTabIndex(isClosed, index)}
            >
              {isClosed ? (
                <>
                  <div className={`rs-base-layer-switcher-title${closedClass}`}>
                    {titles.button}
                  </div>
                  <img
                    src={getNextImage(currentLayer, baseLayers, images)}
                    alt={altText}
                    className="rs-base-layer-switcher-image"
                  />
                </>
              ) : (
                <>
                  <div className="rs-base-layer-switcher-title">
                    {layerName}
                  </div>
                  <img
                    src={images[index]}
                    alt={altText}
                    className="rs-base-layer-switcher-image"
                  />
                </>
              )}
            </div>
          );
        })}
      </div>
    </div>
  );
}

BaseLayerSwitcher.propTypes = propTypes;
BaseLayerSwitcher.defaultProps = defaultProps;

export default BaseLayerSwitcher;
