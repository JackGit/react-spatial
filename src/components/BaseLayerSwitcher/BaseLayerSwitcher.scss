.rs-base-layer-switcher {
  position: relative;
  height: 80px;
  width: 115px;
  transition: 250ms width;
  overflow: hidden;

  &.rs-open {
    width: 600px;
  }

  .rs-base-layer-switcher-layers {
    display: flex;
    overflow: auto;
    height: 100%;
    width: 600px;
  }

  .rs-base-layer-switcher-button {
    margin: 4px;
    background-color: white;
    border: 2px solid white;
  }

  .rs-layer {
    position: relative;
    min-width: 100px;
    overflow: hidden;
    background-color: gray;
  }

  .rs-close {
    display: flex;
    flex-direction: column;
    justify-content: center;
  }

  .rs-open {
    &:hover,
    &.rs-active {
      border: 2px solid red;
    }
  }

  .rs-base-layer-switcher-title {
    position: absolute;
    width: 100%;
    font-size: 10px;
    padding: 2px 0;
    background-color: rgba(255, 255, 255, 0.9);
    text-align: center;
    z-index: 2;

    &.rs-closed {
      color: white;
      background-color: rgba(0, 0, 0, 0.7);
    }
  }

  .rs-base-layer-switcher-image {
    width: 100px;
    height: 100px;
  }
}
