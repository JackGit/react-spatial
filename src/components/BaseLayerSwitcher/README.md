#

This demonstrates the use of BaseLayerSwitcher.

```jsx
import React from 'react';
import OLMap from 'ol/Map';
import BaseLayerSwitcher from 'react-spatial/components/BaseLayerSwitcher';
import BasicMap from 'react-spatial/components/BasicMap';
import LayerService from 'react-spatial/LayerService';
import ConfigReader from 'react-spatial/ConfigReader';

const center = [1149722.7037660484, 6618091.313553318];
const map = new OLMap({ controls: [] });
const layers = ConfigReader.readConfig(treeData).filter(layer => layer.getIsBaseLayer());
const layerService = new LayerService(layers);

<div className="rs-base-layer-example">
  <BasicMap
    map={map}
    center={center}
    zoom={6}
    layers={layers}
  />
  <BaseLayerSwitcher
    layers={layers}
    layerService={layerService}
  />
</div>;
```
