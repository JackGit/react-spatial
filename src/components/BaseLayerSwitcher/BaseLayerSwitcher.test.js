/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';
import renderer from 'react-test-renderer';
import ConfigReader from '../../ConfigReader';
import data from '../../../data/TreeData';
import BaseLayerSwitcher from './BaseLayerSwitcher';

configure({ adapter: new Adapter() });

const getSnapshot = newData => {
  const layers = ConfigReader.readConfig(newData || data);
  const layerImages = {
    layer1: 'foo',
  };
  const component = renderer.create(
    <BaseLayerSwitcher layers={layers} layerImages={layerImages} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

const mountComp = newData => {
  const layers = ConfigReader.readConfig(newData || data);
  const layerImages = {
    layerFoo: 'foo',
    layerBar: 'bar',
    layerFoobar: 'foobar',
  };
  return mount(<BaseLayerSwitcher layers={layers} layerImages={layerImages} />);
};

describe('BaseLayerToggler', () => {
  describe('matches snapshots', () => {
    test('using default properties.', () => {
      getSnapshot();
    });
  });

  test('renders one switch button per baselayer', () => {
    const comp = mountComp(data);
    expect(comp.find('.rs-base-layer-switch-button').length).toBe(3);
  });

  test('the first baselayer is visible on mount', () => {
    const comp = mountComp(data);
    expect(comp.props().layers[0].getVisible()).toBe(true);
  });

  test('unmounts close button and switches layer on click', () => {
    const comp = mountComp(data);
    comp
      .find('.rs-base-layer-switch-button')
      .at(0)
      .simulate('click');
    comp
      .find('.rs-base-layer-switch-button')
      .at(2)
      .simulate('click');
    expect(comp.find('.rs-base-layer-close').exists()).toBe(false);
    expect(
      comp
        .props()
        .layers.filter(layer => layer.getIsBaseLayer())[2]
        .getVisible(),
    ).toBe(true);
  });
});
