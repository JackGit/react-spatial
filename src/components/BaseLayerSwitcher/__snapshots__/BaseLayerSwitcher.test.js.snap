// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLayerToggler matches snapshots using default properties. 1`] = `
<div
  className="rs-base-layer-switcher"
>
  <div
    className="rs-base-layer-switcher-layers"
  >
    <div
      aria-label="Open Baselayer-Switcher"
      className="rs-base-layer-switcher-button rs-layer rs-active"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
      title="Open Baselayer-Switcher"
    >
      <div
        className="rs-base-layer-switcher-title rs-closed"
      >
        Base layers
      </div>
      <img
        alt="Source not found"
        className="rs-base-layer-switcher-image"
      />
    </div>
    <div
      aria-label="Open Baselayer-Switcher"
      className="rs-base-layer-switcher-button rs-layer"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="-1"
      title="Open Baselayer-Switcher"
    >
      <div
        className="rs-base-layer-switcher-title rs-closed"
      >
        Base layers
      </div>
      <img
        alt="Source not found"
        className="rs-base-layer-switcher-image"
      />
    </div>
    <div
      aria-label="Open Baselayer-Switcher"
      className="rs-base-layer-switcher-button rs-layer"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="-1"
      title="Open Baselayer-Switcher"
    >
      <div
        className="rs-base-layer-switcher-title rs-closed"
      >
        Base layers
      </div>
      <img
        alt="Source not found"
        className="rs-base-layer-switcher-image"
      />
    </div>
  </div>
</div>
`;
